include/onnxruntime/core/framework/tensor.h

```diff
@@ -145,6 +145,40 @@ class Tensor final {
   /// <returns>Bytes required.</returns>
   static size_t CalculateTensorStorageSize(MLDataType elt_type, const TensorShape& shape);
 
+  /// <summary>
+  /// Calculate the required storage for the tensor.
+  /// </summary>
+  /// <param name="elt_type">Data type of the tensor elements.</param>
+  /// <param name="shape">Tensor shape.</param>
+  /// <param name="alignment">Power of 2 alignment to include in calculation.
+  /// Bumps up result to the nearest multiple of alignment. Set to 0 to ignore.</param>
+  /// <param name="storage_size">The resulting storage size.</param>
+  /// <returns>Status indicating success or failure.</returns>
+  static Status CalculateTensorStorageSize(MLDataType elt_type, const TensorShape& shape, size_t alignment,
+                                           size_t& storage_size);
+
+  /// <summary>
+  /// Get the number of elements for a Tensor of the given element type and shape.
+  /// For element types smaller than 1 byte (e.g., int4), a single Tensor element stores multiple sub-byte elements.
+  /// So, this function returns the number of Tensor elements, each of which may contain multiple sub-byte elements.
+  /// </summary>
+  /// <param name="elt_type">Data type of the tensor elements.</param>
+  /// <param name="shape">Tensor shape.</param>
+  /// <returns>Number of Tensor elements. Returns -1 if shape has negative dims.</returns>
+  static inline int64_t GetNumTensorElems(MLDataType elt_type, const TensorShape& shape) {
+    return GetNumTensorElems(elt_type, shape.Size());
+  }
+
+  /// <summary>
+  /// Get the number of elements for a Tensor of the given element type and shape size.
+  /// For element types smaller than 1 byte (e.g., int4), a single Tensor element stores multiple sub-byte elements.
+  /// So, this function returns the number of Tensor elements, each of which may contain multiple sub-byte elements.
```

Comment:
I think 'elements' is a confusing name here as onnx.proto and the IR use 'elements' in a way that seems to be 1:1 with what the tensor's shape would define as the size, but the implementation is returning a value 

e.g. int4 is described as a tensor element type 

https://github.com/onnx/onnx/blob/main/docs/IR.md#tensor-element-types
https://github.com/onnx/onnx/blob/b90e252da11dea9bdc191d6b9b8d01511ef3e3bd/onnx/onnx.proto#L564-L565

Would a name more along the lines of NumDataStorageTypeValues clarify? 

Do we need these new functions to be public? Less potential for confusion if they're private/local scope.

Reply:
I agree that is confusing. I do like using something like "tensor data storage type" instead of tensor element.

> Do we need these new functions to be public? Less potential for confusion if they're private/local scope.

These static functions do not strictly need to be public in this PR. However, given that we expose a public `T* Data()` method to get just the pointer to the data, it seems like we should also expose a method to get the count (i.e., see `NumElements()` below which we can rename to `NumDataStorageTypeValues()`).